File: app/api/src/worker.ts
import { type ChildProcess, spawn } from "node:child_process";
import path from "node:path";
import readline from "node:readline";
import { v4 as uuidv4 } from "uuid";
import { getPythonBin, getRepoRoot, pythonEnv } from "@/python";
import type { BatchInferenceParams, InferenceParams, TtsParams } from "@/schemas";

export interface SingleInferRequest {
  command: "infer";
  id: string;
  params: InferenceParams;
  inputPath: string;
  outputPath: string;
  onLog: (msg: string) => void;
  resolve: (res: { outputPath: string; info: string }) => void;
  reject: (err: Error) => void;
}

export interface BatchInferRequest {
  command: "infer_batch";
  id: string;
  params: BatchInferenceParams;
  inputFolder: string;
  outputFolder: string;
  onLog: (msg: string) => void;
  resolve: (res: { info: string }) => void;
  reject: (err: Error) => void;
}

export interface TtsWorkerRequest {
  command: "tts";
  id: string;
  params: Partial<TtsParams>;
  ttsText?: string;
  ttsFile?: string;
  ttsVoice: string;
  ttsRate: number;
  outputTtsPath: string;
  outputRvcPath?: string;
  onLog: (msg: string) => void;
  resolve: (res: { outputTtsPath: string; outputRvcPath?: string; outputPath: string; info: string }) => void;
  reject: (err: Error) => void;
}

export interface AnalyzeAudioRequest {
  command: "analyze_audio";
  id: string;
  inputPath: string;
  plotPath: string;
  onLog: (msg: string) => void;
  resolve: (res: { info: unknown; plot: string }) => void;
  reject: (err: Error) => void;
}

export interface F0CurveRequest {
  command: "f0_curve";
  id: string;
  inputPath: string;
  method: string;
  outputImage: string;
  outputTxt: string;
  onLog: (msg: string) => void;
  resolve: (res: { outputImage: string; outputTxt: string }) => void;
  reject: (err: Error) => void;
}

export interface ModelBlenderRequest {
  command: "model_blender";
  id: string;
  modelName: string;
  pth1: string;
  pth2: string;
  ratio: number;
  onLog: (msg: string) => void;
  resolve: (res: { message: string; file: string | null }) => void;
  reject: (err: Error) => void;
}

export interface InspectModelRequest {
  command: "inspect_model";
  id: string;
  pthPath: string;
  onLog: (msg: string) => void;
  resolve: (res: Record<string, unknown>) => void;
  reject: (err: Error) => void;
}

export type WorkerRequest =
  | SingleInferRequest
  | BatchInferRequest
  | TtsWorkerRequest
  | AnalyzeAudioRequest
  | F0CurveRequest
  | ModelBlenderRequest
  | InspectModelRequest;

export type InferenceRequest = SingleInferRequest;

interface WorkerIPCMessage {
  _applio_ipc?: boolean;
  type: "ready" | "pong" | "log" | "done" | "error" | "unloaded";
  id?: string;
  pid?: number;
  message?: string;
  outputPath?: string;
  outputTtsPath?: string;
  outputRvcPath?: string;
  outputImage?: string;
  outputTxt?: string;
  info?: unknown;
  plot?: string;
  file?: string | null;
  metadata?: Record<string, unknown>;
  error?: string;
  traceback?: string;
}

export class InferenceWorkerManager {
  private child: ChildProcess | null = null;
  private queue: WorkerRequest[] = [];
  private activeJob: WorkerRequest | null = null;
  private isReady = false;
  private readyCallbacks: (() => void)[] = [];
  private restarting = false;
  private warmedUp = false;
  private warmupTimer?: NodeJS.Timeout;

  constructor() {
    process.on("exit", () => this.stop());
  }

  public getPid(): number | undefined {
    return this.child?.pid;
  }

  public start() {
    if (this.child && !this.child.killed) return;
    const repoRoot = getRepoRoot();
    const workerScript = path.join(repoRoot, "rvc", "infer", "worker.py");
    const pyBin = getPythonBin();
    const pathEnv = `${repoRoot}${path.delimiter}${process.env.PATH || ""}`;

    const child = spawn(pyBin, [workerScript], {
      cwd: repoRoot,
      env: pythonEnv({ PATH: pathEnv }),
      windowsHide: true,
      stdio: ["pipe", "pipe", "pipe"],
    });

    this.child = child;
    this.isReady = false;
    this.warmedUp = false;

    if (!child.stdout || !child.stderr) {
      this.child = null;
      return;
    }

    const rlOut = readline.createInterface({ input: child.stdout });
    rlOut.on("line", (line) => {
      const trimmed = line.trim();
      if (!trimmed) return;
      if (trimmed.startsWith('{"_applio_ipc":')) {
        try {
          const msg = JSON.parse(trimmed) as WorkerIPCMessage;
          this.handleIPCMessage(msg);
          return;
        } catch {}
      }
      if (this.activeJob) {
        this.activeJob.onLog(trimmed);
      }
    });

    const rlErr = readline.createInterface({ input: child.stderr });
    rlErr.on("line", (line) => {
      const trimmed = line.trim();
      if (!trimmed) return;
      if (this.activeJob) {
        this.activeJob.onLog(trimmed);
      }
    });

    child.on("error", (err) => {
      this.handleProcessExit(err);
    });

    child.on("exit", (code) => {
      this.handleProcessExit(new Error(`Inference worker exited with code ${code}`));
    });
  }

  private handleIPCMessage(msg: WorkerIPCMessage) {
    if (msg.type === "ready") {
      this.isReady = true;
      const cbs = [...this.readyCallbacks];
      this.readyCallbacks = [];
      for (const cb of cbs) cb();
      this.processNext();
    } else if (msg.type === "log") {
      if (this.activeJob && (!msg.id || this.activeJob.id === msg.id) && msg.message) {
        this.activeJob.onLog(msg.message);
      }
    } else if (msg.type === "done") {
      if (this.activeJob && this.activeJob.id === msg.id) {
        const job = this.activeJob;
        this.activeJob = null;
        if (job.command === "infer") {
          job.resolve({ outputPath: msg.outputPath || "", info: String(msg.info || "") });
        } else if (job.command === "infer_batch") {
          job.resolve({ info: String(msg.info || "Batch inference completed") });
        } else if (job.command === "tts") {
          job.resolve({
            outputTtsPath: msg.outputTtsPath || "",
            outputRvcPath: msg.outputRvcPath,
            outputPath: msg.outputPath || msg.outputRvcPath || msg.outputTtsPath || "",
            info: String(msg.info || "TTS completed"),
          });
        } else if (job.command === "analyze_audio") {
          job.resolve({ info: msg.info, plot: msg.plot || "" });
        } else if (job.command === "f0_curve") {
          job.resolve({ outputImage: msg.outputImage || "", outputTxt: msg.outputTxt || "" });
        } else if (job.command === "model_blender") {
          job.resolve({ message: msg.message || "Model blended", file: msg.file ?? null });
        } else if (job.command === "inspect_model") {
          job.resolve(msg.metadata || {});
        }
        this.processNext();
      }
    } else if (msg.type === "error") {
      if (this.activeJob && this.activeJob.id === msg.id) {
        const job = this.activeJob;
        this.activeJob = null;
        job.reject(new Error(msg.error || "Worker task failed"));
        this.processNext();
      }
    }
  }

  private handleProcessExit(err: Error) {
    this.isReady = false;
    this.child = null;
    if (this.activeJob) {
      const job = this.activeJob;
      this.activeJob = null;
      job.reject(err);
    }
    if (this.queue.length > 0 && !this.restarting) {
      this.restarting = true;
      setTimeout(() => {
        this.restarting = false;
        this.start();
      }, 500);
    }
  }

  public waitReady(): Promise<void> {
    if (this.isReady && this.child) return Promise.resolve();
    this.start();
    return new Promise((resolve) => {
      this.readyCallbacks.push(resolve);
    });
  }

  public warmupDelayed(ms = 3000) {
    if (this.warmupTimer) clearTimeout(this.warmupTimer);
    this.warmupTimer = setTimeout(() => {
      void this.warmup().catch(() => {});
    }, ms);
  }

  public async warmup(): Promise<void> {
    if (this.warmedUp) return;
    await this.waitReady();
    if (this.warmedUp || this.activeJob || this.queue.length > 0 || !this.child) return;
    this.warmedUp = true;
    try {
      this.child.stdin?.write(`${JSON.stringify({ command: "warmup" })}\n`);
    } catch {
      this.warmedUp = false;
    }
  }

  public async preloadModel(pthPath: string, sid = 0): Promise<void> {
    await this.waitReady();
    if (!this.child) return;
    try {
      this.child.stdin?.write(`${JSON.stringify({ command: "preload_model", pthPath, sid })}\n`);
    } catch {
      /* ignore */
    }
  }

  public async infer(
    id: string,
    params: InferenceParams,
    inputPath: string,
    outputPath: string,
    onLog: (msg: string) => void,
  ): Promise<{ outputPath: string; info: string }> {
    return new Promise<{ outputPath: string; info: string }>((resolve, reject) => {
      this.queue.push({
        command: "infer",
        id,
        params,
        inputPath,
        outputPath,
        onLog,
        resolve,
        reject,
      });
      void this.waitReady().then(() => this.processNext());
    });
  }

  public async inferBatch(
    id: string,
    params: BatchInferenceParams,
    inputFolder: string,
    outputFolder: string,
    onLog: (msg: string) => void,
  ): Promise<{ info: string }> {
    return new Promise<{ info: string }>((resolve, reject) => {
      this.queue.push({
        command: "infer_batch",
        id,
        params,
        inputFolder,
        outputFolder,
        onLog,
        resolve,
        reject,
      });
      void this.waitReady().then(() => this.processNext());
    });
  }

  public async tts(
    id: string,
    options: {
      params: Partial<TtsParams>;
      ttsText?: string;
      ttsFile?: string;
      ttsVoice: string;
      ttsRate: number;
      outputTtsPath: string;
      outputRvcPath?: string;
    },
    onLog: (msg: string) => void,
  ): Promise<{ outputTtsPath: string; outputRvcPath?: string; outputPath: string; info: string }> {
    return new Promise<{ outputTtsPath: string; outputRvcPath?: string; outputPath: string; info: string }>(
      (resolve, reject) => {
        this.queue.push({
          command: "tts",
          id,
          params: options.params,
          ttsText: options.ttsText,
          ttsFile: options.ttsFile,
          ttsVoice: options.ttsVoice,
          ttsRate: options.ttsRate,
          outputTtsPath: options.outputTtsPath,
          outputRvcPath: options.outputRvcPath,
          onLog,
          resolve,
          reject,
        });
        void this.waitReady().then(() => this.processNext());
      },
    );
  }

  public async analyzeAudio(
    id: string,
    inputPath: string,
    plotPath: string,
    onLog: (msg: string) => void,
  ): Promise<{ info: unknown; plot: string }> {
    return new Promise<{ info: unknown; plot: string }>((resolve, reject) => {
      this.queue.push({
        command: "analyze_audio",
        id,
        inputPath,
        plotPath,
        onLog,
        resolve,
        reject,
      });
      void this.waitReady().then(() => this.processNext());
    });
  }

  public async extractF0(
    id: string,
    inputPath: string,
    method: string,
    outputImage: string,
    outputTxt: string,
    onLog: (msg: string) => void,
  ): Promise<{ outputImage: string; outputTxt: string }> {
    return new Promise<{ outputImage: string; outputTxt: string }>((resolve, reject) => {
      this.queue.push({
        command: "f0_curve",
        id,
        inputPath,
        method,
        outputImage,
        outputTxt,
        onLog,
        resolve,
        reject,
      });
      void this.waitReady().then(() => this.processNext());
    });
  }

  public async blendModels(
    id: string,
    modelName: string,
    pth1: string,
    pth2: string,
    ratio: number,
    onLog: (msg: string) => void,
  ): Promise<{ message: string; file: string | null }> {
    return new Promise<{ message: string; file: string | null }>((resolve, reject) => {
      this.queue.push({
        command: "model_blender",
        id,
        modelName,
        pth1,
        pth2,
        ratio,
        onLog,
        resolve,
        reject,
      });
      void this.waitReady().then(() => this.processNext());
    });
  }

  public async inspectModel(pthPath: string): Promise<Record<string, unknown>> {
    return new Promise<Record<string, unknown>>((resolve, reject) => {
      this.queue.push({
        command: "inspect_model",
        id: uuidv4(),
        pthPath,
        onLog: () => {},
        resolve,
        reject,
      });
      void this.waitReady().then(() => this.processNext());
    });
  }

  private processNext() {
    if (this.activeJob || this.queue.length === 0 || !this.isReady || !this.child) return;
    const req = this.queue.shift();
    if (!req) return;
    this.activeJob = req;

    let payload: Record<string, unknown>;
    if (req.command === "infer") {
      payload = {
        command: "infer",
        id: req.id,
        inputPath: req.inputPath,
        outputPath: req.outputPath,
        params: req.params,
      };
    } else if (req.command === "infer_batch") {
      payload = {
        command: "infer_batch",
        id: req.id,
        inputFolder: req.inputFolder,
        outputFolder: req.outputFolder,
        params: req.params,
      };
    } else if (req.command === "tts") {
      payload = {
        command: "tts",
        id: req.id,
        ttsText: req.ttsText,
        ttsFile: req.ttsFile,
        ttsVoice: req.ttsVoice,
        ttsRate: req.ttsRate,
        outputTtsPath: req.outputTtsPath,
        outputRvcPath: req.outputRvcPath,
        params: req.params,
      };
    } else if (req.command === "analyze_audio") {
      payload = {
        command: "analyze_audio",
        id: req.id,
        inputPath: req.inputPath,
        plotPath: req.plotPath,
      };
    } else if (req.command === "f0_curve") {
      payload = {
        command: "f0_curve",
        id: req.id,
        inputPath: req.inputPath,
        method: req.method,
        outputImage: req.outputImage,
        outputTxt: req.outputTxt,
      };
    } else if (req.command === "model_blender") {
      payload = {
        command: "model_blender",
        id: req.id,
        modelName: req.modelName,
        pth1: req.pth1,
        pth2: req.pth2,
        ratio: req.ratio,
      };
    } else if (req.command === "inspect_model") {
      payload = {
        command: "inspect_model",
        id: req.id,
        pthPath: req.pthPath,
      };
    } else {
      this.activeJob = null;
      this.processNext();
      return;
    }

    try {
      this.child.stdin?.write(`${JSON.stringify(payload)}\n`);
    } catch (err: unknown) {
      this.activeJob = null;
      req.reject(err instanceof Error ? err : new Error(String(err)));
      this.processNext();
    }
  }

  public stop() {
    if (this.child) {
      try {
        this.child.kill();
      } catch {
        /* ignore */
      }
      this.child = null;
    }
  }
}

export const inferenceWorker = new InferenceWorkerManager();

File: rvc/infer/worker.py
import os
import sys
import json
import traceback

now_dir = os.getcwd()
if now_dir not in sys.path:
    sys.path.append(now_dir)

# Keep a reference to real standard streams before redirection
real_stdout = sys.stdout
real_stderr = sys.stderr

active_job_id = None


def get_active_job_id():
    return active_job_id


class IPCStdoutRedirector:
    def __init__(self, target, job_id_fn):
        self.target = target
        self.job_id_fn = job_id_fn
        self.buffer = ""

    def write(self, s):
        if not s:
            return
        self.buffer += s
        while "\n" in self.buffer:
            line, self.buffer = self.buffer.split("\n", 1)
            line = line.strip()
            if line:
                send_ipc(
                    {
                        "type": "log",
                        "id": self.job_id_fn(),
                        "message": line,
                    }
                )

    def flush(self):
        if self.buffer.strip():
            send_ipc(
                {
                    "type": "log",
                    "id": self.job_id_fn(),
                    "message": self.buffer.strip(),
                }
            )
        self.buffer = ""


def send_ipc(data: dict):
    payload = {"_applio_ipc": True, **data}
    real_stdout.write(json.dumps(payload, ensure_ascii=False) + "\n")
    real_stdout.flush()


def map_params(params: dict, input_path: str, output_path: str) -> dict:
    return {
        "pitch": int(params.get("pitch", 0)),
        "index_rate": float(params.get("indexRate", 0.75)),
        "volume_envelope": float(params.get("volumeEnvelope", 1.0)),
        "protect": float(params.get("protect", 0.5)),
        "f0_method": str(params.get("f0Method", "rmvpe")),
        "input_path": input_path,
        "output_path": output_path,
        "pth_path": params.get("pthPath", ""),
        "index_path": params.get("indexPath", "") or "",
        "split_audio": bool(params.get("splitAudio", False)),
        "f0_autotune": bool(params.get("f0Autotune", False)),
        "f0_autotune_strength": float(params.get("f0AutotuneStrength", 1.0)),
        "proposed_pitch": bool(params.get("proposedPitch", False)),
        "proposed_pitch_threshold": float(params.get("proposedPitchThreshold", 155.0)),
        "clean_audio": bool(params.get("cleanAudio", False)),
        "clean_strength": float(params.get("cleanStrength", 0.5)),
        "export_format": str(params.get("exportFormat", "WAV")),
        "embedder_model": str(params.get("embedderModel", "contentvec")),
        "embedder_model_custom": params.get("embedderModelCustom") or None,
        "formant_shifting": bool(params.get("formantShifting", False)),
        "formant_qfrency": float(params.get("formantQfrency", 1.0)),
        "formant_timbre": float(params.get("formantTimbre", 1.0)),
        "post_process": bool(params.get("postProcess", False)),
        "reverb": bool(params.get("reverb", False)),
        "pitch_shift": bool(params.get("pitchShift", False)),
        "limiter": bool(params.get("limiter", False)),
        "gain": bool(params.get("gain", False)),
        "distortion": bool(params.get("distortion", False)),
        "chorus": bool(params.get("chorus", False)),
        "bitcrush": bool(params.get("bitcrush", False)),
        "clipping": bool(params.get("clipping", False)),
        "compressor": bool(params.get("compressor", False)),
        "delay": bool(params.get("delay", False)),
        "reverb_room_size": float(params.get("reverbRoomSize", 0.5)),
        "reverb_damping": float(params.get("reverbDamping", 0.5)),
        "reverb_wet_gain": float(params.get("reverbWetGain", 0.33)),
        "reverb_dry_gain": float(params.get("reverbDryGain", 0.4)),
        "reverb_width": float(params.get("reverbWidth", 1.0)),
        "reverb_freeze_mode": float(params.get("reverbFreezeMode", 0.0)),
        "pitch_shift_semitones": float(params.get("pitchShiftSemitones", 0.0)),
        "limiter_threshold": float(params.get("limiterThreshold", -6.0)),
        "limiter_release_time": float(params.get("limiterReleaseTime", 0.05)),
        "gain_db": float(params.get("gainDb", 0.0)),
        "distortion_gain": float(params.get("distortionGain", 25.0)),
        "chorus_rate": float(params.get("chorusRate", 1.0)),
        "chorus_depth": float(params.get("chorusDepth", 0.25)),
        "chorus_center_delay": float(params.get("chorusCenterDelay", 7.0)),
        "chorus_feedback": float(params.get("chorusFeedback", 0.0)),
        "chorus_mix": float(params.get("chorusMix", 0.5)),
        "bitcrush_bit_depth": int(params.get("bitcrushBitDepth", 8)),
        "clipping_threshold": float(params.get("clippingThreshold", -6.0)),
        "compressor_threshold": float(params.get("compressorThreshold", 0.0)),
        "compressor_ratio": float(params.get("compressorRatio", 1.0)),
        "compressor_attack": float(params.get("compressorAttack", 1.0)),
        "compressor_release": float(params.get("compressorRelease", 100.0)),
        "delay_seconds": float(params.get("delaySeconds", 0.5)),
        "delay_feedback": float(params.get("delayFeedback", 0.0)),
        "delay_mix": float(params.get("delayMix", 0.5)),
        "sid": int(params.get("sid", 0)),
    }


def run_worker():
    global active_job_id

    # Install redirectors
    sys.stdout = IPCStdoutRedirector(real_stdout, get_active_job_id)
    sys.stderr = IPCStdoutRedirector(real_stdout, get_active_job_id)

    from rvc.infer.infer import VoiceConverter

    # Preload singleton VoiceConverter in background
    vc = VoiceConverter()

    send_ipc({"type": "ready", "pid": os.getpid()})

    for raw_line in sys.stdin:
        line = raw_line.strip()
        if not line:
            continue
        try:
            req = json.loads(line)
        except Exception as e:
            send_ipc({"type": "error", "id": None, "error": f"Invalid JSON: {e}"})
            continue

        cmd = req.get("command")
        job_id = req.get("id")
        active_job_id = job_id

        try:
            if cmd == "ping":
                send_ipc({"type": "pong", "id": job_id})
            elif cmd == "warmup":
                # Warm up CUDA, PyTorch, audio libraries, embedder, and the most recent model
                try:
                    import torch

                    if torch.cuda.is_available():
                        torch.cuda.init()

                    # Pre-import heavy numerical and DSP libraries
                    import librosa
                    import soundfile

                    vc.load_hubert("contentvec", None)

                    # Look for the latest model in logs/ to preload so the first user conversion is instant
                    preloaded_name = None
                    logs_dir = os.path.join(now_dir, "logs")
                    if os.path.isdir(logs_dir):
                        candidates = []
                        for r, _, files in os.walk(logs_dir):
                            for f in files:
                                if f.lower().endswith(".pth") and not f.startswith(
                                    ("G_", "D_")
                                ):
                                    full_p = os.path.join(r, f)
                                    try:
                                        candidates.append(
                                            (os.path.getmtime(full_p), full_p)
                                        )
                                    except Exception:
                                        pass
                        if candidates:
                            candidates.sort(reverse=True)
                            latest_model = candidates[0][1]
                            try:
                                vc.get_vc(latest_model, 0)
                                preloaded_name = os.path.basename(latest_model)
                            except Exception:
                                pass

                    msg = f"Worker warmed up (embedder: contentvec" + (
                        f", active model: {preloaded_name})" if preloaded_name else ")"
                    )
                    send_ipc({"type": "log", "id": job_id, "message": msg})
                except Exception as e:
                    send_ipc(
                        {"type": "log", "id": job_id, "message": f"Warmup notice: {e}"}
                    )
            elif cmd == "unload":
                vc.cleanup_model()
                import torch

                if torch.cuda.is_available():
                    torch.cuda.empty_cache()
                send_ipc({"type": "unloaded", "id": job_id})
            elif cmd == "infer":
                params = req.get("params", {})
                input_path = req.get("inputPath")
                output_path = req.get("outputPath")
                kwargs = map_params(params, input_path, output_path)

                export_format = kwargs.get("export_format", "WAV")
                kwargs["audio_input_path"] = kwargs.pop("input_path", input_path)
                kwargs["audio_output_path"] = kwargs.pop("output_path", output_path)
                kwargs["model_path"] = kwargs.pop("pth_path", "")

                vc.convert_audio(**kwargs)
                final_out = output_path.replace(".wav", f".{export_format.lower()}")
                info_msg = f"File {input_path} inferred successfully."
                sys.stdout.flush()
                send_ipc(
                    {
                        "type": "done",
                        "id": job_id,
                        "success": True,
                        "outputPath": final_out,
                        "info": info_msg,
                    }
                )
            elif cmd == "infer_batch":
                params = req.get("params", {})
                input_folder = req.get("inputFolder")
                output_folder = req.get("outputFolder")
                kwargs = map_params(params, "", "")
                kwargs.pop("input_path", None)
                kwargs.pop("output_path", None)
                kwargs["model_path"] = kwargs.pop("pth_path", "")

                vc.convert_audio_batch(
                    audio_input_paths=input_folder,
                    audio_output_path=output_folder,
                    **kwargs,
                )
                sys.stdout.flush()
                send_ipc(
                    {
                        "type": "done",
                        "id": job_id,
                        "success": True,
                        "info": f"Batch conversion in {input_folder} completed.",
                    }
                )
            elif cmd == "tts":
                import asyncio
                import edge_tts

                tts_text = req.get("ttsText", "")
                tts_file = req.get("ttsFile", "")
                tts_voice = req.get("ttsVoice", "en-US-AnaNeural")
                tts_rate = int(req.get("ttsRate", 0))
                output_tts_path = req.get("outputTtsPath")
                output_rvc_path = req.get("outputRvcPath")
                params = req.get("params", {})

                if tts_file and os.path.exists(tts_file):
                    try:
                        with open(tts_file, "r", encoding="utf-8") as f:
                            tts_text = f.read()
                    except Exception:
                        with open(tts_file, "r") as f:
                            tts_text = f.read()

                if not tts_text.strip():
                    raise ValueError("No text provided for TTS synthesis.")

                os.makedirs(
                    os.path.dirname(os.path.abspath(output_tts_path)), exist_ok=True
                )
                rate_str = f"+{tts_rate}%" if tts_rate >= 0 else f"{tts_rate}%"
                asyncio.run(
                    edge_tts.Communicate(tts_text, tts_voice, rate=rate_str).save(
                        output_tts_path
                    )
                )
                print(f"TTS audio generated at '{output_tts_path}'")

                final_out = output_tts_path
                if output_rvc_path and params.get("pthPath"):
                    os.makedirs(
                        os.path.dirname(os.path.abspath(output_rvc_path)), exist_ok=True
                    )
                    kwargs = map_params(params, output_tts_path, output_rvc_path)
                    kwargs["audio_input_path"] = output_tts_path
                    kwargs["audio_output_path"] = output_rvc_path
                    kwargs["model_path"] = kwargs.pop("pth_path", "")
                    vc.convert_audio(**kwargs)
                    export_format = kwargs.get("export_format", "WAV")
                    final_out = output_rvc_path.replace(
                        ".wav", f".{export_format.lower()}"
                    )
                    print(f"TTS RVC conversion completed at '{final_out}'")

                sys.stdout.flush()
                send_ipc(
                    {
                        "type": "done",
                        "id": job_id,
                        "success": True,
                        "outputTtsPath": output_tts_path,
                        "outputRvcPath": final_out if output_rvc_path else None,
                        "outputPath": final_out,
                        "info": "TTS synthesis completed.",
                    }
                )
            elif cmd == "preload_model":
                pth_path = req.get("pthPath")
                sid = int(req.get("sid", 0))
                if pth_path and os.path.isfile(pth_path):
                    vc.get_vc(pth_path, sid)
                    send_ipc(
                        {
                            "type": "done",
                            "id": job_id,
                            "success": True,
                            "info": f"Model '{os.path.basename(pth_path)}' preloaded.",
                        }
                    )
                else:
                    send_ipc(
                        {
                            "type": "error",
                            "id": job_id,
                            "error": f"Model file not found: {pth_path}",
                        }
                    )
            elif cmd == "inspect_model":
                pth_path = req.get("pthPath")
                if not pth_path or not os.path.isfile(pth_path):
                    raise ValueError(f"Model file not found: {pth_path}")
                from rvc.lib.tools.inspect_checkpoint import inspect_checkpoint

                meta = inspect_checkpoint(pth_path)
                send_ipc(
                    {
                        "type": "done",
                        "id": job_id,
                        "success": True,
                        "metadata": meta,
                    }
                )
            elif cmd == "analyze_audio":
                input_path = req.get("inputPath")
                plot_path = req.get("plotPath")
                if not input_path or not os.path.isfile(input_path):
                    raise ValueError(f"Audio file not found: {input_path}")
                from rvc.lib.tools.analyzer import analyze_audio

                os.makedirs(os.path.dirname(os.path.abspath(plot_path)), exist_ok=True)
                info, plot = analyze_audio(input_path, plot_path)
                send_ipc(
                    {
                        "type": "done",
                        "id": job_id,
                        "success": True,
                        "info": info,
                        "plot": plot,
                    }
                )
            elif cmd == "f0_curve":
                input_path = req.get("inputPath")
                method = req.get("method", "rmvpe")
                output_image = req.get("outputImage")
                output_txt = req.get("outputTxt")
                if not input_path or not os.path.isfile(input_path):
                    raise ValueError(f"Audio file not found: {input_path}")
                from rvc.lib.tools.f0_curve import extract_f0_curve

                os.makedirs(
                    os.path.dirname(os.path.abspath(output_image)), exist_ok=True
                )
                os.makedirs(os.path.dirname(os.path.abspath(output_txt)), exist_ok=True)
                img, txt = extract_f0_curve(
                    input_path, method, output_image, output_txt
                )
                send_ipc(
                    {
                        "type": "done",
                        "id": job_id,
                        "success": True,
                        "outputImage": img,
                        "outputTxt": txt,
                    }
                )
            elif cmd == "model_blender":
                model_name = req.get("modelName")
                pth1 = req.get("pth1")
                pth2 = req.get("pth2")
                ratio = float(req.get("ratio", 0.5))
                from rvc.train.process.model_blender import model_blender

                r = model_blender(model_name, pth1, pth2, ratio)
                msg, f = r if isinstance(r, tuple) else (str(r), None)
                send_ipc(
                    {
                        "type": "done",
                        "id": job_id,
                        "success": True,
                        "message": msg,
                        "file": f,
                    }
                )
            else:
                send_ipc(
                    {
                        "type": "error",
                        "id": job_id,
                        "error": f"Unknown command: {cmd}",
                    }
                )
        except Exception as e:
            sys.stdout.flush()
            tb = traceback.format_exc()
            send_ipc(
                {
                    "type": "error",
                    "id": job_id,
                    "error": str(e),
                    "traceback": tb,
                }
            )
        finally:
            active_job_id = None


if __name__ == "__main__":
    run_worker()
